Topic/montage start end

joeykleingers opened the pull request:
* Adding Montage Start and Montage End parameters to montage import filters.
* Adding Montage Start and Montage End parameters to ITKStitchMontage.
* Updating ITKPCMTileRegistration filter to use IntVec2 instead of IntVec3.

joeykleingers committed:
Adding Montage Start and Montage End parameters to montage import filters.

Parameters are added to ITKImportFijiMontage, ITKImportRobometMontage, ImportZenInfoMontage, and ImportAxioVisionV4Montage.
Changed ITKPCMTileRegistration filter to use IntVec2 parameters.

Signed-off-by: Joey Kleingers <joey.kleingers@bluequartz.net>

joeykleingers committed:
Updating montage end parameter when the cache is reloaded.

Signed-off-by: Joey Kleingers <joey.kleingers@bluequartz.net>

joeykleingers committed:
Fixing incorrect Fiji input file human label and expected extension.

Signed-off-by: Joey Kleingers <joey.kleingers@bluequartz.net>

joeykleingers committed:
Clang format.

Signed-off-by: Joey Kleingers <joey.kleingers@bluequartz.net>

joeykleingers committed:
Revert "Updating montage end parameter when the cache is reloaded."

This reverts commit bddaa00fea89c7c455cdad5dd3409e9d3263e7a6.

joeykleingers committed:
Adding check into montage importer filters to only import tiles within the bounds.

Signed-off-by: Joey Kleingers <joey.kleingers@bluequartz.net>

joeykleingers committed:
WIP: Add Montage start and end to stitching algorithm.

Signed-off-by: Joey Kleingers <joey.kleingers@bluequartz.net>

joeykleingers committed:
Fixing off-by-one bug in stitching filter.

Signed-off-by: Joey Kleingers <joey.kleingers@bluequartz.net>

tplatt77 reviewed (approved):
Looks good to me after review / testing

imikejackson committed:
Improved sanity checks on the montage start/end values

Improved Montage Information String to the users expectations.

Signed-off-by: Michael Jackson <mike.jackson@bluequartz.net>

imikejackson commented:
There are multiple issues with the implementations of the import filters. I updated the ImportAxioVision with the proper expectations. Please update the remaining import filters (Zen, RoboMet, Fiji) to have similar sanity checks, feedback and error checking.

imikejackson commented on ITKImageProcessingFilters/ImportAxioVisionV4Montage.cpp:
```diff
@@ -102,6 +103,8 @@ class ImportAxioVisionV4MontagePrivate
   QString m_MetaDataAttributeMatrixName;
+  int32_t m_MaxRow = 0;
+  int32_t m_MaxCol = 0;
   std::vector<ImportAxioVisionV4Montage::BoundsType> m_BoundsCache;
```
Added a cached value for the max number of columns and rows. Those values can be used when generating the MontageInformation string and also when sanity checking the inputs from the user.

imikejackson commented on ITKImageProcessingFilters/ImportAxioVisionV4Montage.cpp:
```diff
@@ -328,6 +336,43 @@ void ImportAxioVisionV4Montage::dataCheck()
     generateCache(root);
   }
 
+  if(m_MontageStart[0] > m_MontageEnd[0])
```
Added complete sanity checking with verbose error messages to tell the user exactly what went wrong and why the preflight is failing.

imikejackson commented on ITKImageProcessingFilters/ImportAxioVisionV4Montage.cpp:
```diff
@@ -544,7 +591,8 @@ void ImportAxioVisionV4Montage::generateCache(QDomElement& root)
     }
-
+    d_ptr->m_MaxCol = std::max(bound.Col, d_ptr->m_MaxCol);
+    d_ptr->m_MaxRow = std::max(bound.Row, d_ptr->m_MaxRow);
```
Update the cache with the max Column & Row. To be used later in error checking and montage information strings

imikejackson commented on ITKImageProcessingFilters/ImportAxioVisionV4Montage.cpp:
```diff
@@ -595,7 +643,11 @@ void ImportAxioVisionV4Montage::generateCache(QDomElement& root)
   QString montageInfo;
   QTextStream ss(&montageInfo);
-  ss << "Columns=" << m_ColumnCount << "  Rows=" << m_RowCount << "  Num. Images=" << imageCount;
+  ss << "Max Column: " << m_ColumnCount - 1 << "  Max Row: " << m_RowCount - 1 << "  Image Count: " << imageCount;
```
More information added to the Montage Information string which helps the user make the proper decisions based on the max rows/columns in the montage configuration file.

imikejackson commented on ITKImageProcessingFilters/ITKStitchMontage.h:
```diff
@@ -61,7 +60,8 @@ class ITKImageProcessing_EXPORT ITKStitchMontage : public AbstractFilter
   Q_OBJECT
   PYB11_CREATE_BINDINGS(ITKStitchMontage SUPERCLASS AbstractFilter)
-  PYB11_PROPERTY(IntVec3Type MontageSize READ getMontageSize WRITE setMontageSize)
+  PYB11_PROPERTY(IntVec2Type MontageStart READ getMontageStart WRITE setMontageStart)
```
Adding in the Montage Start/End values we no longer need to select any DataContainers because those can now be programmatically generated. This lessens the chance that the user messes up the import. Right now the user must select the start/end values *AND* select the proper data containers to be used. Please update this filter so that the user does not need to select any Data Containers.

imikejackson commented:
All the import filters should be updated to display a montage information much like what is now in the AxioVision Import filter.

The StitchImages filter needs updating relative to the DataContainer selection.

joeykleingers committed:
Implementing suggested fixes.

Signed-off-by: Joey Kleingers <joey.kleingers@bluequartz.net>

joeykleingers committed:
Adding missing Pybind11 macro.

Signed-off-by: Joey Kleingers <joey.kleingers@bluequartz.net>

imikejackson committed:
Rework the Montage Information generation by using a placeholder

The bulk of the message is generated during the generate cache.
When the string is requested the placeholder is replaced with content.

Signed-off-by: Michael Jackson <mike.jackson@bluequartz.net>

imikejackson commented:
double check the error messages. I thought I saw a debug print that basically said there were either too many or not enough arguments to a QString().arg().arg() style of code line.

joeykleingers committed:
Rework the Montage Information generation by using a placeholder.

Adding the work done in ImportAxioVisionV4Montage to the other montage import filters.

Signed-off-by: Joey Kleingers <joey.kleingers@bluequartz.net>

joeykleingers committed:
Fixing QString arg errors.

Signed-off-by: Joey Kleingers <joey.kleingers@bluequartz.net>

imikejackson commented on ITKImageProcessingFilters/ITKStitchMontage.cpp:
```diff
@@ -160,16 +159,46 @@
+QString generateDataContainerName(const QString& dataContainerPrefix, const IntVec2Type& montageStart, const IntVec2Type& montageEnd, int32_t row, int32_t col)
+{
+  IntVec3Type montageSize;
+  std::transform(montageStart.begin(), montageStart.end(), montageEnd.begin(), montageSize.begin(), [](int32_t a, int32_t b) -> int32_t { return a + b + 1; });
```
What is this line supposed to be doing?

joeykleingers replied:
I assume that it computes the size of the montage and stores it in montageSize, using the montageStart and montageEnd?  I grabbed that code from ITKPCMTileRegistration, which it looks like you wrote on 5/23/19.

imikejackson committed:
Rewrite code section to generate the DataContainer Names

Previous version had a bug

Signed-off-by: Michael Jackson <mike.jackson@bluequartz.net>

joeykleingers committed:
Move generateDataContainerName method into MontageImporterHelper.

Signed-off-by: Joey Kleingers <joey.kleingers@bluequartz.net>

joeykleingers committed:
Fixing bugs where montage sizes were not being calculated correctly.

Adding error checks in ITKPCMTileRegistration filter for the montage
start and end values.

Signed-off-by: Joey Kleingers <joey.kleingers@bluequartz.net>

joeykleingers committed:
Fixing crash due to incorrect tile coords passed into ITK.

Signed-off-by: Joey Kleingers <joey.kleingers@bluequartz.net>

joeykleingers committed:
Generating data container name using the common method.

Signed-off-by: Joey Kleingers <joey.kleingers@bluequartz.net>

imikejackson committed:
Properly use itk::Command subclass progress objects.

Signed-off-by: Michael Jackson <mike.jackson@bluequartz.net>
